pilopt/src/lib.rs

```diff
@@ -552,29 +561,151 @@ fn remove_trivial_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
                 }
                 None
             }
-            IdentityKind::Plookup => {
-                assert_eq!(
-                    identity.left.expressions.len(),
-                    identity.right.expressions.len()
-                );
-                identity.left.expressions.is_empty().then_some(index)
+            Identity::Lookup(LookupIdentity { left, right, .. }) => {
+                assert_eq!(left.expressions.len(), right.expressions.len());
+                left.expressions.is_empty().then_some(index)
             }
-            IdentityKind::Permutation => None,
-            IdentityKind::Connect => None,
+            Identity::Permutation(..) => None,
+            Identity::Connect(..) => None,
         })
         .collect();
     pil_file.remove_identities(&to_remove);
 }
 
 fn remove_duplicate_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    // TODO: there must be a less verbose way to do this...
+    /// Wrapper around `Identity` that implements `PartialEq` and `Ord` for canonical comparison, ignoring source information and id.
+    struct CanonicalIdentity<'a, T>(&'a Identity<T>);
+
+    impl<T: PartialEq> PartialEq for CanonicalIdentity<'_, T> {
+        fn eq(&self, other: &Self) -> bool {
```

Comment:
At least this one can be implemented from PartialCmp, right?

Reply:
Excellent point

Reply:
Done using Ord